### PR TITLE
Add account utility functions

### PR DESCRIPTION
**Pull request by CoolNamesAllTaken**
* Modify @cahogan 's `get_balance()` with optional timestamp feature.
* Add account initialization function.
* Add functions for setting / getting interest rate.
* Add function for transferring funds between accounts.
* Add helper function to find the most recent anchor event, or an anchor event just before an optional timestamp.

### CODE REVIEW

File: code/svb-docker/python/svb_web/core/models.py [hunk 1]
```diff
@@ -36,6 +44,9 @@ def save(self, *args, **kwargs):
             # Customer object is being saved for the first time.
             self.customer_id = self.get_customer_id()
         super(Customer, self).save(*args, **kwargs) # call super save function
+    
+    def get_absolute_url(self):
+        return f"c/{self.customer_id}"
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py:49]: Implement c/ url

File: code/svb-docker/python/svb_web/core/models.py [hunk 2]
```diff
@@ -1,7 +1,15 @@
+from __future__ import annotations # allow type hinting with the type of the enclosing class
+
 from django.db import models
-from datetime import date
-import datetime
+from datetime import date, datetime, timezone
 import escpos.printer
+import math
+
+def get_utc_timestamp():
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py:8]:
```suggestion
def get_current_utc_timestamp():

```

File: code/svb-docker/python/svb_web/core/models.py [hunk 3]
```diff
@@ -56,29 +67,210 @@ def save(self, *args, **kwargs):
...
+        @param[in] to_account Account object to transfer funds into.
+        @param[in] amount Amount to transfer.
+        @retval True if success, False if failed.
+        """
+        last_from_account_anchor = from_account.get_last_anchor_event()
+        last_to_account_anchor = to_account.get_last_anchor_event()
+        if last_from_account_anchor is None or last_to_account_anchor is None:
+            raise RuntimeError(f"Attempted a transfer between one or more accounts that weren't initialized.")
+
+        # Create simulataneous deposit and withdrawal anchor events.
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py:208]:
```suggestion
        # Create simultaneous deposit and withdrawal anchor events.

```

File: code/svb-docker/python/svb_web/core/models.py [hunk 4]
```diff
@@ -56,29 +67,210 @@ def save(self, *args, **kwargs):
...
+        transfer_timestamp = datetime.now(tz=timezone.utc)
+        # Create withdrawal anchor event.
+        withdrawal_anchor = AnchorEvent(
+            account=from_account,
+            category=AnchorEvent.WITHDRAWAL,
+            timestamp=transfer_timestamp,
+            balance=from_account.get_balance(timestamp=transfer_timestamp)-amount,
+            # Don't think about the problem case of making a transfer before the most recent AnchorEvent.
+            interest_rate=from_account.get_last_anchor_event().interest_rate
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py:217]:
```suggestion
            interest_rate=last_from_account_anchor.interest_rate

```

File: code/svb-docker/python/svb_web/core/models.py [hunk 5]
```diff
@@ -56,29 +67,210 @@ def save(self, *args, **kwargs):
...
+        )
+        if withdrawal_anchor.balance < 0:
+            print(
+                f"Account transfer of f{amount} treats between account"
+                f"{from_account.account_number} ({from_account.account_name}) ->"
+                f"{to_account.account_number} ({to_account.account_name})"
+                "failed due to insufficinet funds."
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py:224]:
```suggestion
                "failed due to insufficient funds."

```

File: code/svb-docker/python/svb_web/core/models.py [hunk 6]
```diff
@@ -56,29 +67,210 @@ def save(self, *args, **kwargs):
...
+            )
+            return False # abort the transfer due to insufficient funds
+        # Create deposit anchor event.
+        deposit_anchor = AnchorEvent(
+            account=to_account,
+            category=AnchorEvent.DEPOSIT,
+            timestamp=transfer_timestamp,
+            balance=to_account.get_balance(timestamp=transfer_timestamp)+amount,
+            # Don't think about the problem case of making a transfer before the most recent AnchorEvent.
+            interest_rate=to_account.get_last_anchor_event().interest_rate
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py:234]:
```suggestion
            interest_rate=last_to_account_anchor.interest_rate

```

File: code/svb-docker/python/svb_web/core/models.py [hunk 7]
```diff
@@ -56,29 +67,210 @@ def save(self, *args, **kwargs):
...
     
 class AnchorEvent(models.Model):
     # Anchor Events:
     # - Account created.
     # - Interest rate changed.
     # - Candy deposited into account.
     # - Candy withdrawn from account.
+    CATEGORY_NAME_MAX_LENGTH = 20
+
+    ACCOUNT_CREATED = "UPDATE_INTEREST"
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py:253]: Is this a typo?
```suggestion
    ACCOUNT_CREATED = "ACCOUNT_CREATED"

```

File: code/svb-docker/python/svb_web/core/models.py.orig [hunk 8]
```diff
@@ -0,0 +1,177 @@
+from django.db import models
+<<<<<<< HEAD
+from datetime import date, datetime
+=======
+from datetime import date
+import datetime
+>>>>>>> main
+import escpos.printer
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py.orig:8]:
```suggestion
from django.db import models
from datetime import date, datetime
import escpos.printer

```

File: code/svb-docker/python/svb_web/core/models.py.orig
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/models.py.orig:1]: Delete file?

File: affinity/svb_debit_card/exports/svb_logo_silouhette.png
Comment (cahogan) [inline affinity/svb_debit_card/exports/svb_logo_silouhette.png:1]: File has typo in name: silhouette

File: code/svb-docker/python/svb_web/svb_db_init.yaml
Comment (cahogan) [inline code/svb-docker/python/svb_web/svb_db_init.yaml:1]: You should think about adding initial anchorevents too?

Comment (cahogan) [reply code/svb-docker/python/svb_web/svb_db_init.yaml:1]: how and when is this called?

File: code/svb-docker/python/svb_web/svb_web/__pycache__/__init__.cpython-310.pyc
Comment (cahogan) [inline code/svb-docker/python/svb_web/svb_web/__pycache__/__init__.cpython-310.pyc:1]: Stop tracking .pyc files

File: code/svb-docker/python/svb_web/core/views/internal/teller.py
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/views/internal/teller.py:1]: Is this view used?

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
+import pytest, time
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:1]:
```suggestion
import pytest
import time

```

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    # Test that account numbers are incrementing properly.
+    assert account1.account_number < account2.account_number
+    # Test that accounts can be retrieved by name.
+    assert Account.objects.filter(account_name__exact="test1")[0] == account1
+    # Test that accounts can be deleted.
+    account1.delete()
+    assert len(Account.objects.filter(account_name__exact="test1")) == 0
+    
+    # Reset database to original state.
+    account2.delete()
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:35]: The db decorator already does this for you!

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+@pytest.mark.django_db
+def test_account_create_delete():
+    customer = Customer(
+        customer_id="test_cust",
+        first_name="frank",
+        costume="lyboring",
+        referrer=None,
+        security_candy="toots"
+    )
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:14]:
```suggestion

```

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+def test_account_init_delete():
+    account1 = Account(
+        account_name="test1",
+        customer=None
+    )
+    account1.save()
+    # Account not yet initialized, there should be no associated anchor events.
+    assert len(AnchorEvent.objects.filter(account__exact=account1)) == 0
+    
+    # Test that account can't be initialized in the future.
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:47]: best to separate these into multiple tests

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        account1.init()
+    
+    # Delete the initialization AnchorEvent and init with new parameters.
+    init_anchor_event.delete()
+    assert len(AnchorEvent.objects.filter(pk__exact=init_anchor_event.pk))==0
+    # Should be able to init again now that first anchor event is gone.
+    init_timestamp=datetime.now(tz=timezone.utc)
+    init_anchor_event = account1.init(
+        timestamp=init_timestamp,
+        balance=100.3,
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:65]: Why the magic number?

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        interest_rate=3
+    )
+    matched_init_anchor_events = AnchorEvent.objects.filter(account__exact=account1)
+    assert len(matched_init_anchor_events) == 1
+    retrieved_init_anchor_event = matched_init_anchor_events[0]
+    assert(retrieved_init_anchor_event.timestamp == init_anchor_event.timestamp)
+    # Cast these both to floats since Balance is a Decimal to avoid floating point error
+    # and now it bite us in the butt.
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:73]:
```suggestion
    # and it is still the right choice for a legitimate financial institution like ours.

```

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        test_account.get_interest_rate()
+    # Also can't set interest rate before the account is initiated.
+    with pytest.raises(RuntimeError):
+        test_account.set_interest_rate(9.0)
+    test_account.init()
+    test_account.set_interest_rate(5.3)
+    assert test_account.get_interest_rate() == 5.3
+    saved_timestamp = datetime.now(tz=timezone.utc) # save this for later
+    # Set and get a new interest rate.
+    # FIXME: Known limitation, if we update interest rates with intervals faster than 100ms, we exceed the
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:96]: 👀

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+    # resolution of our timestamps and get Anchor Event out of order errors (graceful failure, not hidden).
+    time.sleep(0.1)
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:98]: 😞

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/tests/test_account_model.py:1]: Ideally, split these tests up a lot.

**Review by cahogan — COMMENTED** [review]
Clean up this PRs extraneous files, then I'll approve. Cool work!

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py (hunk 16 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:98]: This is fixed with the anchor event overwriting functionality!

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py (hunk 14 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:73]: lmao

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py (hunk 13 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:65]: For testing!

File: code/svb-docker/python/svb_web/core/models.py (hunk 2 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/models.py:8]: Resolved.

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:1]: no

File: code/svb-docker/python/svb_web/core/migrations/0015_account_account_name_alter_anchorevent_category.py
[the remark this answers is not in the document]
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/migrations/0015_account_account_name_alter_anchorevent_category.py:1]: image
help

File: code/svb-docker/python/svb_web/core/migrations/0015_account_account_name_alter_anchorevent_category_and_more.py [hunk 17]
```diff
@@ -1,23 +1,29 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='account',
+            name='account_name',
+            field=models.CharField(default='', max_length=80),
+        ),
+        migrations.AlterField(
+            model_name='anchorevent',
+            name='category',
+            field=models.CharField(choices=[('UPDATE_INTEREST', 'Create the account.'), ('DEPOSIT', 'Deposit funds into the associated account.'), ('WITHDRAWAL', 'Withdraw funds from the associated account.'), ('UPDATE_INTEREST', 'Update the interest rate of the associated account.')], default=None, max_length=20),
```
Comment (cahogan) [inline code/svb-docker/python/svb_web/core/migrations/0015_account_account_name_alter_anchorevent_category_and_more.py:22]: Fix your choices: UPDATE_INTEREST -> CREATE_ACCOUNT

**Review by cahogan — CHANGES_REQUESTED** [review]

File: affinity/svb_debit_card/exports/svb_logo_silouhette.png
Comment (CoolNamesAllTaken) [reply affinity/svb_debit_card/exports/svb_logo_silouhette.png:1]: sharp eyes

like a hawk

File: code/svb-docker/python/svb_web/svb_web/__pycache__/__init__.cpython-310.pyc
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/svb_web/__pycache__/__init__.cpython-310.pyc:1]: Just updated it, thanks

File: code/svb-docker/python/svb_web/core/views/internal/teller.py
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/views/internal/teller.py:1]: Not yet aaaaa

File: code/svb-docker/python/svb_web/core/forms.py [hunk 18]
```diff
@@ -41,7 +41,44 @@ def clean(self):
                 )
         return cleaned_data
 
+class AccountTransactionForm(forms.Form):
```
[the remark this answers is not in the document]
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/forms.py:44]: it's a helpful guide

File: code/svb-docker/python/svb_web/svb_db_init.yaml
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/svb_db_init.yaml:1]: u will call this

Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/svb_db_init.yaml:1]: A gift from me to you <3

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py (hunk 10 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:37]: cooooooool

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py (hunk 11 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:16]: Why u change this?

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py (hunk 12 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:46]: good feedback but no time gogogo

File: code/svb-docker/python/svb_web/core/tests/test_account_model.py (hunk 11 above)
Comment (CoolNamesAllTaken) [reply code/svb-docker/python/svb_web/core/tests/test_account_model.py:16]: Oh I see not used

**Pull request merged**